### PR TITLE
feat: split builder into more refined modules

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [x] New feature (a non-breaking change which adds functionality)

## Description

- [x] @nuxt/generator
- [x] @nuxt/webpack
- [x] refactor coupling between @nuxt/builder and @nuxt/webpack

## Checklist:

- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/builder/src/builders/webpack.js [hunk 1]
```diff
@@ -0,0 +1,297 @@
+import path from 'path'
```
Comment (pi0) [inline packages/builder/src/builders/webpack.js]: @clarkdo Should we keep this file into `builder` package or still a part of WIP?

Comment (clarkdo) [reply packages/builder/src/builders/webpack.js]: It’s not finished yet 😊

Comment (pi0) [reply packages/builder/src/builders/webpack.js]: Forgive me. I'm too excited to see the progress 💃

File: packages/webpack/src/builder.js [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+          name,
+          compiler,
+          stats
+        })
+
+        // Reload renderer if available
+        this.nuxt.renderer.loadResources(this.mfs || fs)
+
+        // Resolve on next tick
+        process.nextTick(resolve)
```
Comment (pi0) [inline packages/webpack/src/builder.js]: As we probably lose history, would you please add a hint comment to the original issue?

File: packages/webpack/src/builder.js [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    if (this.nuxt.renderer) {
+      this.nuxt.renderer.webpackDevMiddleware = this.webpackDevMiddleware
+      this.nuxt.renderer.webpackHotMiddleware = this.webpackHotMiddleware
+    }
+
+    // Start watching client files
+    this.watchClient()
+  }
+
+  watchClient() {
```
Comment (pi0) [inline packages/webpack/src/builder.js]: Watch utilities are completely decoupled from webpack. What's your idea to move them into base builder? So we can reuse it for other builder :) @clarkdo

File: packages/webpack/src/builder.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+  r,
+  parallel,
+  sequence
+} from '@nuxt/common'
+
+import { ClientConfig, ServerConfig, PerfLoader } from './config'
+
+export default class WebpackBuilder {
+  constructor(context) {
+    this.context = context
```
Comment (pi0) [inline packages/webpack/src/builder.js]: For easier tracing and allowing context references to be modified on runtime, suggest just keep using `this.context` 👍

File: packages/common/src/build/context.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export default class BuildContext {
+  constructor(builder) {
+    this.nuxt = builder.nuxt
```
Comment (pi0) [inline packages/common/src/build/context.js]: Idea stolen from @clarkdo :P We can do `Object.assign(this, DEFAULTS, builder)`

**Review by pi0 — DISMISSED** [review]

File: packages/webpack/src/builder.js (hunk 4 above)
Comment (clarkdo) [reply packages/webpack/src/builder.js]: Cool

File: packages/webpack/src/builder.js (hunk 3 above)
Comment (clarkdo) [reply packages/webpack/src/builder.js]: Opps, it was in base builder, my mistake.

File: packages/common/src/build/context.js (hunk 5 above)
Comment (clarkdo) [reply packages/common/src/build/context.js]: I thought about this, it may assign all properties in builder, I feel it's too much 😄

**Comment by clarkdo** [comment]
@pi0 Please review 😸

**Review by pi0 — DISMISSED** [review]
Nice and Neat!

**Review by galvez — DISMISSED** [review]
Awesome :)

**Comment by pi0** [comment]
@clark @galvez Would you please resolve test/util conflict? Not sure most imports should be kept or not 🤔

**Comment by clarkdo** [comment]
Working on it.

**Review by pi0 — APPROVED** [review]

**Pull request merged**